src/main/java/com/google/sps/servlets/BlobServlet.java

```diff
@@ -33,59 +41,203 @@
+    }
+
+    // Asset to update must already exist
+    if (isMask) {
+      Query imgQuery = new Query("Image").setAncestor(projKey);
+      Filter imgFilter =
+          new FilterPredicate("name", FilterOperator.EQUAL, parentImg);
+      imgQuery.setFilter(imgFilter);
+      PreparedQuery existingImg = datastore.prepare(imgQuery);
+
+      if (existingImg.countEntities() == 0) {
+        throw new IOException(
+            "No parent image with that name exists for this project or this project does not exist.");
+      }
+      if (mode == "create") {
+        Key parentEntityKey =
+            existingImg.asSingleEntity().getKey();
+        imgEntity = new Entity("Mask", parentEntityKey);
+      } else {
+        assetParentKey = existingImg.asSingleEntity().getKey();
+
+        Query maskQuery = new Query("Mask").setAncestor(assetParentKey);
+        Filter maskFilter =
+            new FilterPredicate("name", FilterOperator.EQUAL, imgName);
+        imgQuery.setFilter(maskFilter);
+        PreparedQuery existingMask = datastore.prepare(maskQuery);
+
+        if (existingMask.countEntities() == 0) {
+          throw new IOException(
+              "No mask with that name exists to update for this project.");
+        }
+        imgEntity = existingMask.asSingleEntity();
+      }
+    } else if (mode == "update") {
+      Query imgQuery = new Query("Image").setAncestor(projKey);
+      Filter imgFilter =
+          new FilterPredicate("name", FilterOperator.EQUAL, imgName);
+      imgQuery.setFilter(imgFilter);
+      PreparedQuery existingImg = datastore.prepare(imgQuery);
+
+      if (existingImg.countEntities() == 0) {
+        throw new IOException(
+            "No image with that name exists to update for this project or this project does not exist.");
+      }
+      imgEntity = existingImg.asSingleEntity();
+    }
+
+    // Owners have additional permissions
+    String ownersString =
+        (String)accessibleProjects.asSingleEntity().getProperty("owners");
+    Boolean isOwner = ownersString.contains(uid);
+
+    Boolean delete = Boolean.parseBoolean(request.getParameter("delete"));
+    if (delete) {
+      if (!isOwner || mode == "create") {
+        throw new IOException(
+            "Only owners can delete assets and only in update mode.");
+      } else {
+        datastore.delete(imgEntity.getKey());
+        response.sendRedirect("/");
+        return;
+      }
+    }
+
+    if (mode == "create" && !isOwner && !isMask) {
+      throw new IOException("Only owners can add new images.");
+    }
 
     BlobstoreService blobstoreService =
         BlobstoreServiceFactory.getBlobstoreService();
     Map<String, List<BlobKey>> blobs = blobstoreService.getUploads(request);
     List<BlobKey> blobKeys = blobs.get("image");
 
     // User submitted form without selecting a file
-    if (blobKeys == null || blobKeys.isEmpty()) {
-      return;
+    Boolean hasNonEmptyImage = !(blobKeys == null || blobKeys.isEmpty());
```

Comment:
I'd suggest the non negative form.

`blobKeys != null && !blobKeys.IsEmpty()`

Reply:
Noted